File: android_app/app/src/main/java/com/oceansentinels/app/data/remote/dto/IncidentDto.kt
package com.oceansentinels.app.data.remote.dto

import com.google.gson.annotations.SerializedName
import com.oceansentinels.app.domain.model.*
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Incident DTOs for API communication
 */

data class IncidentDto(
    @SerializedName("id") val id: Int,
    @SerializedName("reference_id") val referenceId: String,
    @SerializedName("hazard_type") val hazardType: String,
    @SerializedName("location") val location: String,
    @SerializedName("latitude") val latitude: Double?,
    @SerializedName("longitude") val longitude: Double?,
    @SerializedName("description") val description: String,
    @SerializedName("urgency") val urgency: String,
    @SerializedName("status") val status: String,
    @SerializedName("reporter_id") val reporterId: Int,
    @SerializedName("verified_by_id") val verifiedById: Int?,
    @SerializedName("photo_url") val photoUrl: String?,
    @SerializedName("contact_info") val contactInfo: String?,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("updated_at") val updatedAt: String?,
    @SerializedName("verified_at") val verifiedAt: String?,
    @SerializedName("resolved_at") val resolvedAt: String?,
    @SerializedName("assigned_to_id") val assignedToId: Int? = null,
    @SerializedName("assigned_at") val assignedAt: String? = null
) {
    fun toDomain(): Incident {
        return Incident(
            id = id,
            referenceId = referenceId,
            hazardType = HazardType.fromValue(hazardType),
            location = location,
            latitude = latitude,
            longitude = longitude,
            description = description,
            urgency = UrgencyLevel.fromValue(urgency),
            status = IncidentStatus.fromValue(status),
            reporterId = reporterId,
            verifiedById = verifiedById,
            photoUrl = photoUrl,
            contactInfo = contactInfo,
            createdAt = createdAt.parseDateTime() ?: LocalDateTime.now(),
            updatedAt = updatedAt?.parseDateTime(),
            verifiedAt = verifiedAt?.parseDateTime(),
            resolvedAt = resolvedAt?.parseDateTime(),
            assignedToId = assignedToId,
            assignedAt = assignedAt?.parseDateTime()
        )
    }
}

data class IncidentListResponseDto(
    @SerializedName("incidents") val incidents: List<IncidentDto>,
    @SerializedName("total") val total: Int,
    @SerializedName("page") val page: Int,
    @SerializedName("size") val size: Int,
    @SerializedName("has_next") val hasNext: Boolean,
    @SerializedName("has_prev") val hasPrev: Boolean
) {
    fun toDomain(): IncidentListResult {
        return IncidentListResult(
            incidents = incidents.map { it.toDomain() },
            total = total,
            page = page,
            size = size,
            hasNext = hasNext,
            hasPrev = hasPrev
        )
    }
}

data class CreateIncidentRequestDto(
    @SerializedName("hazard_type") val hazardType: String,
    @SerializedName("location") val location: String,
    @SerializedName("latitude") val latitude: Double?,
    @SerializedName("longitude") val longitude: Double?,
    @SerializedName("description") val description: String,
    @SerializedName("urgency") val urgency: String,
    @SerializedName("contact_info") val contactInfo: String?,
    @SerializedName("mesh_message_id") val meshMessageId: String? = null
) {
    companion object {
        fun fromDomain(request: CreateIncidentRequest): CreateIncidentRequestDto {
            return CreateIncidentRequestDto(
                hazardType = request.hazardType.value,
                location = request.location,
                latitude = request.latitude,
                longitude = request.longitude,
                description = request.description,
                urgency = request.urgency.value,
                contactInfo = request.contactInfo
            )
        }
    }
}

// Extension function for parsing date strings
private fun String.parseDateTime(): LocalDateTime? {
    return try {
        LocalDateTime.parse(this, DateTimeFormatter.ISO_DATE_TIME)
    } catch (e: Exception) {
        try {
            LocalDateTime.parse(this.replace("Z", ""), DateTimeFormatter.ISO_LOCAL_DATE_TIME)
        } catch (e: Exception) {
            null
        }
    }
}

/**
 * Request DTO for bulk mesh message status check.
 * Sent when a device comes online to discover which queued messages
 * have already been uploaded to the server by other mesh devices.
 */
data class MeshCheckRequestDto(
    @SerializedName("message_ids") val messageIds: List<String>
)

/**
 * Response DTO for bulk mesh message status check.
 * 'delivered' = already in the server DB (can stop relaying)
 * 'unknown' = not yet on server (keep relaying or deliver ourselves)
 */
data class MeshCheckResponseDto(
    @SerializedName("delivered") val delivered: List<String>,
    @SerializedName("unknown") val unknown: List<String>
)

File: IOS/OceanSentinels/Data/Remote/DTO/IncidentDTO.swift
import Foundation

// MARK: - Incident DTOs

struct IncidentDTO: Codable {
    let id: Int
    let referenceId: String
    let hazardType: String
    let location: String
    let latitude: Double?
    let longitude: Double?
    let description: String
    let urgency: String
    let status: String
    let reporterId: Int
    let verifiedById: Int?
    let photoUrl: String?
    let contactInfo: String?
    let createdAt: String
    let updatedAt: String?
    let verifiedAt: String?
    let resolvedAt: String?
    let assignedToId: Int?
    let assignedAt: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case referenceId = "reference_id"
        case hazardType = "hazard_type"
        case location, latitude, longitude, description, urgency, status
        case reporterId = "reporter_id"
        case verifiedById = "verified_by_id"
        case photoUrl = "photo_url"
        case contactInfo = "contact_info"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case verifiedAt = "verified_at"
        case resolvedAt = "resolved_at"
        case assignedToId = "assigned_to_id"
        case assignedAt = "assigned_at"
    }
    
    func toDomain() -> Incident {
        Incident(
            id: id,
            referenceId: referenceId,
            hazardType: HazardType.fromValue(hazardType),
            location: location,
            latitude: latitude,
            longitude: longitude,
            description: description,
            urgency: UrgencyLevel.fromValue(urgency),
            status: IncidentStatus.fromValue(status),
            reporterId: reporterId,
            verifiedById: verifiedById,
            photoUrl: photoUrl,
            contactInfo: contactInfo,
            createdAt: createdAt.parseISO8601() ?? Date(),
            updatedAt: updatedAt?.parseISO8601(),
            verifiedAt: verifiedAt?.parseISO8601(),
            resolvedAt: resolvedAt?.parseISO8601(),
            assignedToId: assignedToId,
            assignedAt: assignedAt?.parseISO8601()
        )
    }
}

struct IncidentListResponseDTO: Codable {
    let incidents: [IncidentDTO]
    let total: Int
    let page: Int
    let size: Int
    let hasNext: Bool
    let hasPrev: Bool
    
    enum CodingKeys: String, CodingKey {
        case incidents, total, page, size
        case hasNext = "has_next"
        case hasPrev = "has_prev"
    }
    
    func toDomain() -> IncidentListResult {
        IncidentListResult(
            incidents: incidents.map { $0.toDomain() },
            total: total,
            page: page,
            size: size,
            hasNext: hasNext,
            hasPrev: hasPrev
        )
    }
}

struct CreateIncidentRequestDTO: Codable {
    let hazardType: String
    let location: String
    let latitude: Double?
    let longitude: Double?
    let description: String
    let urgency: String
    let contactInfo: String?
    let meshMessageId: String?
    
    enum CodingKeys: String, CodingKey {
        case hazardType = "hazard_type"
        case location, latitude, longitude, description, urgency
        case contactInfo = "contact_info"
        case meshMessageId = "mesh_message_id"
    }
    
    static func fromDomain(_ request: CreateIncidentRequest) -> CreateIncidentRequestDTO {
        CreateIncidentRequestDTO(
            hazardType: request.hazardType.rawValue,
            location: request.location,
            latitude: request.latitude,
            longitude: request.longitude,
            description: request.description,
            urgency: request.urgency.rawValue,
            contactInfo: request.contactInfo,
            meshMessageId: nil
        )
    }
}

struct MessageResponseDTO: Codable {
    let message: String
}

struct MeshCheckRequestDTO: Codable {
    let messageIds: [String]
    
    enum CodingKeys: String, CodingKey {
        case messageIds = "message_ids"
    }
}

struct MeshCheckResponseDTO: Codable {
    let delivered: [String]
    let unknown: [String]
}

// MARK: - Date Parsing Helper

extension String {
    func parseISO8601() -> Date? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: self) { return date }
        
        formatter.formatOptions = [.withInternetDateTime]
        if let date = formatter.date(from: self) { return date }
        
        // Fallback: remove trailing Z and parse as local
        let cleaned = self.replacingOccurrences(of: "Z", with: "")
        let df = DateFormatter()
        df.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
        df.locale = Locale(identifier: "en_US_POSIX")
        return df.date(from: cleaned)
    }
}
